assembl/auth/make_saml.py

```diff
@@ -0,0 +1,109 @@
+        key_size=2048,
+        backend=default_backend()
+    )
+    key_text = key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption())
+    return (cleanup_x509_text(key_text), key)
+
+
+def private_key_from_cleaned_text(key_text):
+    key_text = '\n'.join([l.strip() for l in
+                          str(key_text).strip('\n').split('\n')])
+    key_text = "-----BEGIN RSA PRIVATE KEY-----\n\n%s\n-----END RSA PRIVATE KEY-----\n" % key_text
+    return serialization.load_pem_private_key(
+        key_text, password=None, backend=default_backend())
+
+
+def make_saml_cert(key, country='', state='', locality='', org='', cn='',
+                   email='', alt_names=None, days=365):
+    if isinstance(key, (str, unicode)):
+        key = private_key_from_cleaned_text(key)
+    if alt_names and isinstance(alt_names(str, unicode)):
+        alt_names = alt_names.split()
+    subject = x509.Name([
+        # Provide various details about who we are.
+        x509.NameAttribute(NameOID.COUNTRY_NAME, unicode(country)),
+        x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, unicode(state)),
+        x509.NameAttribute(NameOID.LOCALITY_NAME, unicode(locality)),
+        x509.NameAttribute(NameOID.ORGANIZATION_NAME, unicode(org)),
+        x509.NameAttribute(NameOID.COMMON_NAME, unicode(cn)),
+        x509.NameAttribute(NameOID.EMAIL_ADDRESS, unicode(email)),
+    ])
+    pkey = key.public_key()
+    skid = x509.SubjectKeyIdentifier.from_public_key(pkey)
+    # Create self-signed
+    serial_number = random_serial_number()
+    crt = x509.CertificateBuilder().subject_name(
+        subject
+    ).issuer_name(
+        subject
+    ).public_key(
+        pkey
+    ).serial_number(
+        serial_number
+    ).not_valid_before(
+        datetime.datetime.utcnow()
+    ).not_valid_after(
+        # Our certificate will be valid for 10 days
+        datetime.datetime.utcnow() + datetime.timedelta(days=days)
+    ).add_extension(
+        x509.BasicConstraints(True, 0), False
+    ).add_extension(
+        skid, False
+    )
+    crt = crt.add_extension(
+        x509.AuthorityKeyIdentifier(
+            skid.digest, [x509.DirectoryName(subject)], serial_number), False
+    )
+    if alt_names:
+        # Describe what sites we want this certificate for.
+        crt = crt.add_extension(
+            x509.SubjectAlternativeName([
+                x509.DNSName(n) for n in altnames]),
+            critical=False)
+    crt = crt.sign(key, hashes.SHA256(), default_backend())
+    crt_text = crt.public_bytes(serialization.Encoding.PEM)
+    return (cleanup_x509_text(crt_text), crt)
+
+
+if __name__ == '__main__':
+    saml_info = {"country": "FR", "state": "Hauts-de-Seine",
+                 "locality": "Levallois-Perret", "org": "Bluenove",
+                 "cn": "assembl-enterprise.bluenove.com",
```

Comment:
again, this code will not be exercized IRL. I have built this information into bluenove.rc, in the private config repository, because that applies to all bluenove servers.